### PR TITLE
ON-46912 # Added types for custom pdfs

### PR DESCRIPTION
**Pull request by jdawg093**
## Requester Checklist

Please only check the items that you have actioned. Do not check items that are not applicable to your PR.

### Implementation

- [ ] Have you tested your implementation locally?
- [ ] If applicable, has an appropriate changelog entry been added? Do not include if you are fixing/changing something that is only in the current release.
- [ ] There are no warnings that have been suppressed unnecessarily
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Have all OneBlink dependency updates been completed (eg apps / apps-react / types etc).
- [ ] Have you ensured this change does not add unwanted dependencies?
- [ ] If this PR contains a refactor, have relevant Jira testing tasks added?
- [ ] Changes that will knowingly make the feature/bug incomplete have been commented with TODO and a description of what needs to be done to finish the feature/bug
- [ ] Any changes made to public APIs have been reflected in the documentation
- [ ] Have you isolated business logic where possible to allow for unit testing?

### Logging and Debugging

- [ ] Are the error messages, if any, informative?
- [ ] Are there enough log events and are they written in a way that allows for easy debugging?
- [ ] "Debugging" code removed
- [ ] Front-end: No erroneous `Console.WriteLines`

### Readability

- [ ] All class, variable, property and method modifiers are provided with the smallest scope possible
- [ ] New files, variables and functions are descriptive/comprehensible and named consistently.
- [ ] There is no dead code (unreachable code)
- [ ] There is no usage of [magic numbers](<https://en.wikipedia.org/wiki/Magic_number_(programming)>)
- [ ] There is no commented out code.
- [ ] In hard-to-understand areas, comments exist and describe rationale or reasons for decisions in code

### Security

- [ ] All personal data inputs are checked (for the correct type, length/size, format, and range).
- [ ] No sensitive information is logged or visible in a stacktrace
- [ ] Are authorization and authentication handled correctly?
- [ ] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [ ] Is data retrieved from external APIs or libraries checked for security issues?
- [ ] Do API endpoints return appropriate status codes

## Reviewer Guide

- It is important that you understand the purpose of the PR.
- You are encouraged to engage with the requestor if you do not understand any of the proposed code changes/additions/deletions.
- Do you, the reviewer, understand what the code does? Do you think a specific expert, like a security expert or a usability expert, should look over the code before it can be accepted?
- Is a framework, API, library, or service used that should not be used? Are there alternatives you could recommend?
- Are there existing hooks/components/functions in the same code base that could be utilised?
- When reviewing tests, attempt to identify missing edge cases that may be relevant to the proposed implementation
- Ensure you check for:
  - Security
  - Scalability
  - Performance
  - Maintainability

### CODE REVIEW

File: typescript/submissionEvents.d.ts [hunk 1]
```diff
@@ -84,6 +84,19 @@ export type PdfSubmissionEventEmailTemplateMapping = {
...
 )
+
+export type BaseCustomPDFFieldMapping = {
+  /** The name of the field that will be replaced in the custom PDF. */
+  replaceableField: string
+}
+
+export type CustomPDFConfiguration = {
+  /** The URL of the PDF that will be used as a template. */
+  pdfUrl: string
```
Comment (mymattcarroll) [inline typescript/submissionEvents.d.ts:95]: Is the intention to store this file publicly?

Prob better to store it as an `S3Configuration` object.

Comment (Zaxist) [reply typescript/submissionEvents.d.ts:95]: Also need a list of stored PDFs on the form level.

Comment (mymattcarroll) [reply typescript/submissionEvents.d.ts:95]: Ah, true. This should prob be an id to reference one of those.

File: typescript/forms.d.ts [hunk 2]
```diff
@@ -1120,8 +1129,11 @@ export type NewForm = {
    * from the console
    */
   isArchived?: boolean
+  /** The custom PDFs that are associated with the form that can be used in place of an OOTB PDF. */
+  customPDFs: FormCustomPDF[]
```
Comment (mymattcarroll) [inline typescript/forms.d.ts:1133]:
```suggestion
  customPDFs?: FormCustomPDF[]
```

File: typescript/forms.d.ts [hunk 3]
```diff
@@ -927,6 +928,14 @@ export type ExternalIdGeneration =
   | EndpointConfiguration
   | ExternalIdGenerationReceiptId
 
+  
+export type FormCustomPDF = {
+  /** The Id of the custom PDF. */
+  id: number
+  /** The S3 Configuration of the custom PDF.  */
+  pdfUrl: S3Configuration
```
Comment (mymattcarroll) [inline typescript/forms.d.ts:936]:
```suggestion
  s3: S3Configuration
```

File: typescript/forms.d.ts [hunk 4]
```diff
@@ -927,6 +928,14 @@ export type ExternalIdGeneration =
   | EndpointConfiguration
   | ExternalIdGenerationReceiptId
 
+  
+export type FormCustomPDF = {
+  /** The Id of the custom PDF. */
+  id: number
```
Comment (mymattcarroll) [inline typescript/forms.d.ts:934]:
```suggestion
  id: string
```

better off making this a UUID

File: typescript/forms.d.ts [hunk 5]
```diff
@@ -927,6 +928,14 @@ export type ExternalIdGeneration =
   | EndpointConfiguration
   | ExternalIdGenerationReceiptId
 
+  
+export type FormCustomPDF = {
```
Comment (mymattcarroll) [inline typescript/forms.d.ts:932]: Prob want to include a label/name property as well.

**Review by mymattcarroll — APPROVED** [review]
